Check arguments of `MissingIndicator` imputer when handling sparse arrays

btel opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Fixes issue #12750 

#### What does this implement/fix? Explain your changes.

This PR implements a simple check of the arguments to make sure that the MissingIndicator raises an exception when both arguments `sparse=True` and `missing_values=0` or array is sparse and `missing_values=0`. It fixes old tests that allowed for these cases and implements one new test.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

btel committed:
adding argument check for the case spare=True and missing_values=0
issue #12750

btel committed:
missing indicator: handle case array is sparse and missing values==0
issue #12750

jeremiedbb commented on sklearn/tests/test_impute.py:
```diff
@@ -934,6 +937,7 @@ def test_missing_indicator_error(X_fit, X_trans, params, msg_err):
         indicator.fit(X_fit).transform(X_trans)
 
 
+
```
Remove this line

glemaitre replied:
Remove this line

jeremiedbb commented on sklearn/impute.py:
```diff
@@ -1109,6 +1109,7 @@ class MissingIndicator(BaseEstimator, TransformerMixin):
 
     def __init__(self, missing_values=np.nan, features="missing-only",
                  sparse="auto", error_on_new=True):
+
```
You don't need to add this blank line

glemaitre replied:
Could you remove this line

glemaitre commented on sklearn/impute.py:
```diff
@@ -1207,6 +1208,17 @@ def fit(self, X, y=None):
                              "'all'. Got {} instead.".format(self.features))
 
+        if isinstance(self.sparse, bool) and self.sparse and self.missing_values == 0:
+            raise ValueError("'missing_values' can not be 0 when 'sparse' is True")
```
```suggestion
            raise ValueError("'missing_values' cannot be 0 when 'sparse' is True")
```

btel replied:
Ok, fixed.

glemaitre commented on sklearn/tests/test_impute.py:
```diff
@@ -965,6 +969,13 @@ def test_missing_indicator_new(missing_values, arr_type, dtype, param_features,
     indicator = MissingIndicator(missing_values=missing_values,
                                  features=param_features,
                                  sparse=False)
+
```
Remove this line

btel committed:
fix flake8 issues

glemaitre commented on sklearn/tests/test_impute.py:
```diff
@@ -1009,9 +1025,22 @@ def test_missing_indicator_sparse_param(arr_type, missing_values,
     indicator = MissingIndicator(missing_values=missing_values,
                                  sparse=param_sparse)
+
+    if param_sparse is True and missing_values == 0:
```
```suggestion
    if param_sparse and missing_values == 0:
```

btel replied:
wouldn't it also cover the case when "param_sparse=='auto'", which should be an allowed case

glemaitre commented on sklearn/tests/test_impute.py:
```diff
@@ -1009,9 +1025,22 @@ def test_missing_indicator_sparse_param(arr_type, missing_values,
     X_fit_mask = indicator.fit_transform(X_fit)
     X_trans_mask = indicator.transform(X_trans)
 
+
```
remove on of the line

btel replied:
ok, I removed all extra blank lines. Thanks!

jeremiedbb commented on sklearn/impute.py:
```diff
@@ -1207,6 +1208,17 @@ def fit(self, X, y=None):
 
+        if isinstance(self.sparse, bool) and self.sparse and self.missing_values == 0:
+            raise ValueError("'missing_values' can not be 0 when 'sparse' is True")
+
```
I don't think we want this case to raise an error.

btel replied:
ok, I removed this case. Thanks!

glemaitre reviewed (changes requested):
Could you add an entry in what's new

jeremiedbb commented on sklearn/impute.py:
```diff
@@ -1207,6 +1208,17 @@ def fit(self, X, y=None):
+
+        if sparse.issparse(X):
+            # missing_values = 0 not allowed with sparse data as it would
+            # force densification
```
You also need to raise the same error in `transform`

jeremiedbb commented on sklearn/tests/test_impute.py:
```diff
@@ -923,6 +923,9 @@ def test_iterative_imputer_early_stopping():
       "'sparse' has to be a boolean or 'auto'"),
+     (np.array([[-1, 1], [1, 2]]), np.array([[-1, 1], [1, 2]]),
+      {'missing_values': 0, 'sparse': True},
+      "'missing_values' can not be 0 when 'sparse' is True"),
```
We should accept this case

jnothman commented on sklearn/impute.py:
```diff
@@ -1207,6 +1208,18 @@ def fit(self, X, y=None):
             raise ValueError("'features' has to be either 'missing-only' or "
                              "'all'. Got {} instead.".format(self.features))
 
+        if self.sparse is True and self.missing_values == 0:
```
I don't think the sparsity of the *output* is the problem, only that of the *input*.

btel replied:
Ok, so in this case we will pass a sparse array without densifying it with missing_value=0 (which is implicit in sparse array)?

jeremiedbb replied:
The output is only ones and zeros. It has nothing to do with the value of `missing_values`. We only want to raise an error when the input is a sparse matrix and missing values == 0

jeremiedbb commented on sklearn/tests/test_impute.py:
```diff
@@ -1009,9 +1025,22 @@ def test_missing_indicator_sparse_param(arr_type, missing_values,
+    if param_sparse is True and missing_values == 0:
+        with pytest.raises(ValueError):
+            X_fit_mask = indicator.fit_transform(X_fit)
+        return
```
We should accept this case

btel replied:
ok, done!

btel committed:
remove empty lines

btel committed:
remove sparse==True and missing_value==0 case; handle `transform` method

btel committed:
remove blankline

btel commented:
ok. I fixed all the issues that came up in the review. Most importantly,

* removed spourius blank lines
* removed the check for arguments `sparse=True` and `missing_values=0`, which, as pointed out by @jnothman, was not necessary
* added check in `transform` method

jeremiedbb commented:
I think it's fine now. But maybe it would be more readable to separate the case sparse input + missing = 0, i.e. have a test which checks everything works on normal inputs and another one which checks that an error is raise otherwise.

btel commented:
@jeremiedbb that makes sense, but I would need to skip over some input combinations.  Otherwise, I couldn't use purest decorators. How would you do that?

glemaitre commented:
I means that we need to specified the combination in a single parametrize instead of creating several of them and make the product. 



If there is no too much case I agree with jerimie
                                                                                                                                      
 

                            
     

Sent from my phone - sorry to be brief and potential misspell.

jeremiedbb commented:
I would remove the missing=0 from the current test and add a test with missing=0.

btel committed:
separate out test for missing_value==0 and sparse input

btel commented:
ok. I enumerated all possible combinations not involving missing_values == 0 and sparse inputs. For the latter case, I added a new test.

btel committed:
remove spurious blank lines

btel committed:
added what's new entry

btel committed:
fixing lint issues

jeremiedbb commented on doc/whats_new/v0.21.rst:
```diff
@@ -133,6 +133,10 @@ Support for Python 3.4 and below has been officially dropped.
+  
+- |Fix| In :class:`impute.MissingIndicator` avoid implicit densification by
+  raising an exception if input is sparse add `missing_values` property
+  is set to 0. :issue:`12750` by :user:`Bartosz Telenczuk <btel>`.
```
Counter intuitive but the `issue` number should be your PR number :)

btel replied:
Fixed. Thanks!

jeremiedbb commented on doc/whats_new/v0.21.rst:
```diff
@@ -167,7 +171,7 @@ Support for Python 3.4 and below has been officially dropped.
 - |Fix| Fixed a bug in :class:`linear_model.LassoLarsIC`, where user input
    ``copy_X=False`` at instance creation would be overridden by default
-   parameter value ``copy_X=True`` in ``fit``. 
+   parameter value ``copy_X=True`` in ``fit``.
```
Please avoid fixing pep8 issues not related to your PR. It makes reviews harder to follow

btel replied:
Ok. I will avoid it for future PRs, but I guess I should keep this one.

jeremiedbb replied:
we usually ask contributors to revert it :/

jeremiedbb commented on sklearn/impute.py:
```diff
@@ -1207,6 +1207,14 @@ def fit(self, X, y=None):
+            # force densification
+            if self.missing_values == 0:
+                raise ValueError("Imputation not possible when missing_values "
+                                 "== 0 and input is sparse. Provide a dense "
```
It's not Imputation here. Maybe something like "Sparse input with missing_values=0 is not supported."

btel replied:
Ok, I changed it.

jeremiedbb commented on sklearn/impute.py:
```diff
@@ -1207,6 +1207,14 @@ def fit(self, X, y=None):
                              "'all'. Got {} instead.".format(self.features))
 
+        if sparse.issparse(X):
+            # missing_values = 0 not allowed with sparse data as it would
```
Since `fit` and `transform` both call `_validate_imput`, we can move this test into this function to avoid redundant code.

btel replied:
Good idea. Thanks!

jeremiedbb commented on sklearn/tests/test_impute.py:
```diff
@@ -992,11 +999,40 @@ def test_missing_indicator_new(missing_values, arr_type, dtype, param_features,
+        indicator.fit_transform(X_fit)
+    with pytest.raises(ValueError):
+        indicator.transform(X_trans)
+
```
I'm not sure the second part raises the right error. I think the error raised is a non fitted estimator error since the fit should have failed above. To be sure, you can add a `match` parameter to `pytest.raises`, with a part of the error message you want to catch. 

You'll then need to fit it on proper data to check the transform.

btel replied:
Good catch! As you said, it was raising "instance not fitted" exception instead of "sparse with missing_values=0". I fixed the test.

btel committed:
remove redundant code

btel committed:
fix issue number in "what's new"

btel committed:
move duplicate code to _validate_input

btel committed:
fix test for transform method

btel committed:
remove nested if conditions

jeremiedbb reviewed (approved):
LGTM

glemaitre commented:
LGTM. Will merged when CI are green.

agramfort commented:
thx @btel

btel commented:
my first every scikit-learn PR merged :confetti_ball: Thanks @jeremiedbb @glemaitre @jnothman @agramfort  for help!

glemaitre commented:
:beer: Thanks a lot.
